File: edgequake/sdks/typescript/src/resources/costs.ts
/**
 * Costs resource — cost tracking, history, budgets, and estimation.
 *
 * WHY: Updated to match Rust costs_types.rs exactly.
 * Adds pricing, estimation, and workspace cost summary endpoints.
 *
 * @module resources/costs
 * @see edgequake/crates/edgequake-api/src/handlers/costs.rs
 */

import type {
  AvailablePricingResponse,
  BudgetInfo,
  CostHistoryQuery,
  CostHistoryResponse,
  CostSummaryResponse,
  EstimateCostRequest,
  EstimateCostResponse,
  UpdateBudgetRequest,
  WorkspaceCostSummaryResponse,
} from "../types/costs.js";
import { Resource } from "./base.js";

export class CostsResource extends Resource {
  /** Get cost summary for the workspace. */
  async summary(): Promise<CostSummaryResponse> {
    return this._get("/api/v1/costs/summary");
  }

  /** Get cost history over time. */
  async history(query?: CostHistoryQuery): Promise<CostHistoryResponse> {
    const params = new URLSearchParams();
    if (query?.start_date) params.set("start_date", query.start_date);
    if (query?.end_date) params.set("end_date", query.end_date);
    if (query?.granularity) params.set("granularity", query.granularity);
    const qs = params.toString();
    return this._get(`/api/v1/costs/history${qs ? `?${qs}` : ""}`);
  }

  /** Get current budget status. */
  async budget(): Promise<BudgetInfo> {
    return this._get("/api/v1/costs/budget");
  }

  /** Update budget settings. */
  async updateBudget(request: UpdateBudgetRequest): Promise<BudgetInfo> {
    return this._patch("/api/v1/costs/budget", request);
  }

  /** Get available model pricing configurations. */
  async pricing(): Promise<AvailablePricingResponse> {
    return this._get("/api/v1/costs/pricing");
  }

  /** Estimate cost for a given model and token counts. */
  async estimate(request: EstimateCostRequest): Promise<EstimateCostResponse> {
    return this._post("/api/v1/costs/estimate", request);
  }

  /** Get workspace cost summary (includes per-operation breakdown). */
  async workspaceSummary(): Promise<WorkspaceCostSummaryResponse> {
    return this._get("/api/v1/costs/workspace");
  }
}

File: edgequake/sdks/rust/src/resources/costs.rs
//! Costs resource.

use crate::client::EdgeQuakeClient;
use crate::error::Result;
use crate::types::operations::*;

pub struct CostsResource<'a> {
    pub(crate) client: &'a EdgeQuakeClient,
}

impl<'a> CostsResource<'a> {
    /// `GET /api/v1/costs/summary`
    pub async fn summary(&self) -> Result<CostSummary> {
        self.client.get("/api/v1/costs/summary").await
    }

    /// `GET /api/v1/costs/history`
    pub async fn history(&self) -> Result<Vec<CostEntry>> {
        self.client.get("/api/v1/costs/history").await
    }

    /// `GET /api/v1/costs/budget`
    pub async fn budget(&self) -> Result<BudgetInfo> {
        self.client.get("/api/v1/costs/budget").await
    }

    /// `PATCH /api/v1/costs/budget` — Update budget settings.
    pub async fn update_budget(&self, body: &serde_json::Value) -> Result<BudgetInfo> {
        self.client
            .patch("/api/v1/costs/budget", Some(body))
            .await
    }

    /// `GET /api/v1/pipeline/costs/pricing` — Get model pricing.
    pub async fn pricing(&self) -> Result<serde_json::Value> {
        self.client.get("/api/v1/pipeline/costs/pricing").await
    }

    /// `POST /api/v1/pipeline/costs/estimate` — Estimate cost for a workload.
    pub async fn estimate(&self, body: &serde_json::Value) -> Result<serde_json::Value> {
        self.client
            .post("/api/v1/pipeline/costs/estimate", Some(body))
            .await
    }
}
